### PR TITLE
New language features

### PR DESCRIPTION
**Pull request by drewnoakes**
Follows #4507, #4548.

Changes proposed in this pull request:
 - Use newer C# language constructs where appropriate, including
   - pattern matching
   - local functions
   - default parameter values
   - object/collection initialisers
   - null coalescence/propagation

The intent is that these changes improve one or more of: terseness, readability, maintainability, performance. Any changes that reviewers don't feel improve any of these enough to warrant the change can be reverted.

As before, one kind of change per commit. Probably easiest to review one commit at a time.
 
What did I do to test the code and ensure quality:
 - Manual review of changes (largely automated)
 - Unit tests
 - Manual testing

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

**Comment by sharwell** [comment]
I prefer to not take 4ceca87, or at minimum to deal with it as a separate change.

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -1311,12 +1311,14 @@ public bool ExistsMergeCommit(string startRev, string endRev)
...
             string revisions = RunGitCmd("rev-list --parents --no-walk " + startRev + ".." + endRev);
             string[] revisionsTab = revisions.Split('\n');
-            Func<string, bool> ex = (string parents) =>
-                {
-                    string[] tab = parents.Split(' ');
-                    return tab.Length > 2 && tab.All(parent => GitRevision.Sha1HashRegex.IsMatch(parent));
-                };
-            return revisionsTab.Any(ex);
+
+            bool IsTwoSha1Hashes(string parents)
```
Comment (sharwell) [inline GitCommands/Git/GitModule.cs]: :bulb: We typically place local functions at the end of the method, after the last `return`. For a `void`-returning method, we would include an explicit `return` statement prior to the first local function.

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 2]
```diff
@@ -1283,17 +1286,16 @@ private void UnstageAllToolStripMenuItemClick(object sender, EventArgs e)
...
-            stageAreaLoaded = () =>
+            void StageAreaLoaded()
             {
                 _currentFilesList = Unstaged;
                 RestoreSelectedFiles(Unstaged.GitItemStatuses, Staged.GitItemStatuses, lastSelection);
                 Unstaged.Focus();
 
-                OnStageAreaLoaded -= stageAreaLoaded;
-            };
+                OnStageAreaLoaded -= StageAreaLoaded;
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormCommit.cs]: ❗️ This won't work. The delegate used for registering and unregistering the event is no longer the same. All event registrations that switched to local functions should be reverted.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs [hunk 3]
```diff
@@ -286,29 +286,25 @@ public void Clear()
...
-                return;
-            }
-
             if (_repository == null)
             {
                 return;
             }
 
-            foreach (RepositoryCategory newRepositoryCategory in Repositories.RepositoryCategories)
+            if (sender is ToolStripItem toolStripItem)
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: ❕  Use the following instead:

```csharp
if (!(sender is ToolStripItem toolStripItem))
{
    return;
}
```

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs [hunk 4]
```diff
@@ -49,7 +49,7 @@ public DashboardCategory(string title, RepositoryCategory repositoryCategory)
 
         public RepositoryCategory RepositoryCategory
         {
-            get { return _repositoryCategory; }
+            get => _repositoryCategory;
```
Comment (sharwell) [inline GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: 👎 I'm not a fan of mixing expression-bodied accessors and block-bodied accessors in the same property.

File: GitExtensions/Program.cs [hunk 5]
```diff
@@ -204,8 +204,7 @@ private static void HandleConfigurationException(System.Configuration.Configurat
                                 if (DialogResult.OK.Equals(MessageBox.Show(string.Format("Files have been deleted.{0}{0}Would you like to attempt to restart GitExtensions?", Environment.NewLine), "Configuration Error", MessageBoxButtons.OKCancel, MessageBoxIcon.Information)))
                                 {
                                     var args = Environment.GetCommandLineArgs();
-                                    System.Diagnostics.Process p = new System.Diagnostics.Process();
-                                    p.StartInfo.FileName = args[0];
+                                    var p = new System.Diagnostics.Process { StartInfo = { FileName = args[0] } };
```
Comment (sharwell) [inline GitExtensions/Program.cs]: 💭 I don't see how this is an improvement

File: GitCommands/Patch/PatchManager.cs [hunk 6]
```diff
@@ -226,10 +226,11 @@ internal class PatchLine
 
         public PatchLine Clone()
         {
-            var c = new PatchLine();
-            c.Text = Text;
-            c.Selected = Selected;
-            return c;
+            return new PatchLine
```
Comment (sharwell) [inline GitCommands/Patch/PatchManager.cs]: ❓ Why would we not want to convert this to a constructor instead?

File: GitUI/CommandsDialogs/FormFormatPatch.cs [hunk 7]
```diff
@@ -2,6 +2,8 @@
 using System.IO;
 using System.Net;
 using System.Net.Mail;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
```
Comment (sharwell) [inline GitUI/CommandsDialogs/FormFormatPatch.cs]: :bulb: Likely an unnecessary addition

File: GitCommands/Git/GitModule.cs (hunk 1 above)
Comment (drewnoakes) [reply GitCommands/Git/GitModule.cs]: I do this too (though I hadn't considered the `return` in `void` methods, which is nice). However moving them will increase the diff considerably which I've had push back on in previous reviews. Can make this change if needed.

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 2 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/FormCommit.cs]: Are you sure about this? I've seen the bug you're talking about but I cannot trigger it in a scenario like this in a unit test. Here is a simple test that shows the same scenario passing:

```c#
private event Action ActionEvent;

[Test]
public void LocalFunctionLambdaTest()
{
    Assert.Null(ActionEvent);

    ActionEvent += Handler;

    Assert.NotNull(ActionEvent);

    ActionEvent();

    Assert.Null(ActionEvent);

    void Handler()
    {
        ActionEvent -= Handler;
    }
}
```

I've tried introducing closures and other delegate signatures but without success. My guess is the compiler closes over the reference to the instance of the local function. Will play with this some more, but so far I can't find a need to revert anything here.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 3 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: I used to bail early on type tests in the way you're suggesting (via `as` then `null` test), but with this new language feature I find the negative pattern harder to read, and even though `toolStripItem` is in scope outside of the block, I prefer to see it used within the block. A matter of taste I guess.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 4 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: I'm not so fussed. To me they're different members and it's less egregious than say an if/else block where one has braces and the other doesn't. Again, a matter of taste. I find the arrow and lack of `return` easier on the eyes.

File: GitExtensions/Program.cs (hunk 5 above)
Comment (drewnoakes) [reply GitExtensions/Program.cs]: Again a matter of taste. I like construction as an expression where possible, and hope event subscriptions can be added to object initialisers one day. I've seen some discussion about this on csharplang.

File: GitCommands/Patch/PatchManager.cs (hunk 6 above)
Comment (drewnoakes) [reply GitCommands/Patch/PatchManager.cs]: I've added constructor `PatchLine(string text, bool selected = false)`. We cannot make the properties get-only, though `Text` can have a private setter.

File: GitUI/CommandsDialogs/FormFormatPatch.cs (hunk 7 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/FormFormatPatch.cs]: Yes. Not sure how that got there. Will remove.

**Comment by drewnoakes** [comment]
> I prefer to not take 4ceca87, or at minimum to deal with it as a separate change.

Would you please explain why? Genuinely curious.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 3 above)
Comment (sharwell) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: I understand that the syntax is somewhat of a preference, but between a return-early style and the new pattern matching syntax I would much rather keep the return-early style.

File: GitCommands/Git/GitModule.cs (hunk 1 above)
Comment (RussKie) [reply GitCommands/Git/GitModule.cs]: If this change is done as a separate PR - we can possibly get it quickly even though the diff may be destructive.
We have to be conscious though of other pending PRs - some of them are quite large (#3693 #4499 etc) and rebasing them is a royal pain

**Comment by sharwell** [comment]
> Would you please explain why? Genuinely curious.

There are different issues that apply to different places, which is why I didn't want them to dominate this PR.

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 2 above)
Comment (sharwell) [reply GitUI/CommandsDialogs/FormCommit.cs]: 👍 I believe you are correct. The big problem occurs when the code relies on two delegates having the same form:

```csharp
Event += (sender, e) => { };
Event -= (sender, e) => { }; // May or may not unregister
```

If this code, you only have one definition of the functions.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 3 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: In many other cases I totally agree with you. This one feels different to me -- for now. It's a newer language feature. Anyway, I had a commit on my machine that addressed this very issue in a few other places (unrelated to pattern matching) which I just submitted #4622.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 4 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: Agree with Sam - I prefer not to mix, looks inconsistent

**Comment by vbjay** [comment]
Make them both lambdas.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 4 above)
Comment (drewnoakes) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: Fixed in 1887d7134.

**Comment by drewnoakes** [comment]
> Make them both lambdas

In this case the setter has two statements and must have statement form.

The issue at hand is formatting of this:

```c#
public RepositoryCategory RepositoryCategory
{
    get { return _repositoryCategory; }
    set
    {
        _repositoryCategory = value;
        InitRepositoryCategory();
    }
}
```

...vs this...

```c#
public RepositoryCategory RepositoryCategory
{
    get => _repositoryCategory;
    set
    {
        _repositoryCategory = value;
        InitRepositoryCategory();
    }
}
```

**Comment by vbjay** [comment]
I see. Didn't see the that in the diff.

**Comment by drewnoakes** [comment]
Anything needed here?

**Comment by RussKie** [comment]
I will have a look at it in the next few days.

File: GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs (hunk 3 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/BrowseDialog/DashboardControl/DashboardCategory.cs]: this proposal kind of negates already merged #4622

**Comment by RussKie** [comment]
Sorry Drew, but I'm finding b5468f5396a9b6bcbcc9da570a9568ebf6325055 to make property definitions inconsistent. Whilst some properties can be fully expressed as `=>`, most can't be.
1887d7134b53366b94b10670fbf24f07dbe99524 reverts only one instance of many, so to save effort I'd rather revert b5468f5396a9b6bcbcc9da570a9568ebf6325055 altogether.
Going forward we can leave it up to the contributors' discretion/subject to reviews.

Other than that I'm happy to accept the proposed changes.

**Comment by RussKie** [comment]
Sorry Drew, but I'm finding b5468f5396a9b6bcbcc9da570a9568ebf6325055 to make property definitions inconsistent. Whilst some properties can be fully expressed as `=>`, most can't be.
1887d7134b53366b94b10670fbf24f07dbe99524 reverts only one instance of many, so to save effort I'd rather revert b5468f5396a9b6bcbcc9da570a9568ebf6325055 altogether.
Going forward we can leave it up to the contributors' discretion/subject to reviews.

Other than that I'm happy to accept the proposed changes.

**Comment by drewnoakes** [comment]
Fixed and rebased. I kept just those few cases where both getter and setter were single expressions (122caf3).

**Pull request merged**